### DISCUSSION TITLE
JTAG/SWD locking nrf52840

### DISCUSSION THREAD

**Discussion by lawrencek52** [pyocd/pyOCD#1215] [Q&A] [answered]
The documentation in https://github.com/pyocd/pyOCD/blob/main/docs/security.md does a great job of describing how to unlock a chip if it is locked (although it seems a little out of date, some of the sample commands returned errors). But security.md fails to describe how to *lock* a chip if it is unlocked.

Can someone help me discover the correct incantation to lock a chip please?

**Comment by flit** [comment, marked as answer]
Hi @lawrencek52, to lock an nRF52 you need to program the `APPROTECT` field of the UICR region of flash. In nRF52840_PS_v1.2.pdf this is §4.5.1.5 on page 44. 

To do this with pyocd:

1. Create a 0x400 byte (1 KiB) binary file containing all 0xFF bytes.
2. Set the byte at offset 0x208 to 0x00. This changes the `PALL` field of  `APPROTECT` to enable debug port protection.
3. Use pyocd to load this binary file to base address 0x10001000, the start of UICR: `pyocd load -a 0x10001000 my_uicr.bin`.
3. After resetting the device (not sure if a cold reset is required), you should no longer be able to connect to the device without getting a "No cores found" warning.

Please note that I haven't tried locking an nRF52 myself, I've only tested writing to the UICR.

(Thanks for mentioning the command errors in the security docs.)

**Reply by lawrencek52** [reply]
Thank you Chris @flit . I did know that I had to write to the UICR-APPROTECT register, it just wasn't clear how this was done with PyOCD. Your instructions are clear. I'll test them in the next week or two and let you know if there is any difficulty.
